### PR TITLE
GH-3102 favour BindingSetAssignment if it binds variables in the query

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #3102 

Briefly describe the changes proposed in this PR:

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

**Comment by hmottestad** [comment]
needs more tests

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryJoinOptimizer.java [hunk 1]
```diff
@@ -439,6 +443,23 @@ protected double getTupleExprCost(TupleExpr tupleExpr, Map<TupleExpr, Double> ca
 				}
 			}
 
+			// BindingSetAssignment has a typical constant cost. This cost is not based on statistics so is much more
+			// reliable. If the BindingSetAssignment binds to any of the other variables in the other tuple expressions
+			// to choose from, then the cost of the BindingSetAssignment should be set to 0 since it will always limit
+			// the upper bound of any other costs. This way the BindingSetAssignment will be chosen as the left
+			// argument.
+			if (tupleExpr instanceof BindingSetAssignment) {
```
Comment (jeenbroekstra) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryJoinOptimizer.java:451]: I get the idea but this should be solved in the EvaluationStatistics implementation. In fact, its default implement currently just assigns BindingSetAssignment a fixed cost of 1. Are you seeing any real cases where that leads to a BindingSetAssignment _not_ being optimally planned?

Comment (jeenbroekstra) [reply core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryJoinOptimizer.java:451]: Ah, hang on, I see you explained this on the issue ticket, l'll read

**Comment by hmottestad** [comment]
@jeenbroekstra I've added a test to this and I'm happy with the code. I've targeted this against 4.0.0 instead of a bug fix release.

**Review by jeenbroekstra — APPROVED** [review]

**Pull request merged**